File: 0823-binary-trees-with-factors/0823-binary-trees-with-factors.java
class Solution {
    public int numFactoredBinaryTrees(int[] arr) {
        long cnt = 0L, mod = (long)1e9+7;
        Arrays.sort(arr);
        HashMap<Integer, Long> dp = new HashMap<>();
        for(int i=0; i<arr.length; i++){
            dp.put(arr[i], 1L);
            for(int j=0; j<i; j++){
                if(arr[i]%arr[j] == 0){
                    dp.put(arr[i], (dp.get(arr[i]) + dp.get(arr[j]) * dp.getOrDefault(arr[i]/arr[j], 0L)) % mod);
                }
            }
            cnt = (cnt + dp.get(arr[i])) % mod;
        }
        return (int)cnt;
    }
}

File: 0823-binary-trees-with-factors/0823-binary-trees-with-factors.cpp
class Solution {
public:
    int numFactoredBinaryTrees(vector<int>& arr) {
        int cnt = 0, mod = 1e9+7;
        unordered_map<int, long> dp;
        sort(arr.begin(), arr.end());
        for(int i=0; i<arr.size(); i++){
            dp[arr[i]] = 1;
            for(int j=0; j<i; j++){
                if(arr[i]%arr[j] == 0){
                    dp[arr[i]] = (dp[arr[i]] + dp[arr[j]] * dp[arr[i]/arr[j]]) % mod;
                }
            }
            cnt = (cnt + dp[arr[i]]) % mod;
        }
        return cnt;
    }
};